File: legacy/bff/src/services/formance.ts
import { SDK, SDKOptions } from "@formance/formance-sdk";
import {
  AccountType,
  Payment,
  PaymentScheme,
  PaymentStatus,
  PaymentType,
} from "@formance/formance-sdk/sdk/models/shared";
import { config } from "../../config";
import { CustomLogger } from "../services/logger";
import { CreatePaymentResponse } from "@formance/formance-sdk/sdk/models/operations";

const logger = new CustomLogger("formance", "service");

const formanceBaseURL = config.formanceBaseURL;
const formanceRequireLogin = config.formanceRequireLogin;
const formanceClientId = config.formanceClientId;
const formanceClientSecret = config.formanceClientSecret;
const formanceConnectorId = config.formanceConnectorId;

let formanceSDKOptions: SDKOptions;
if (formanceRequireLogin) {
  formanceSDKOptions = {
    serverURL: formanceBaseURL,
    security: {
      clientID: formanceClientId,
      clientSecret: formanceClientSecret,
    },
    timeoutMs: 60000,
  };
} else {
  formanceSDKOptions = {
    serverURL: formanceBaseURL,
    timeoutMs: 60000,
  };
}
const formanceSDK: SDK = new SDK(formanceSDKOptions);

export const findPayment = async (referenceId: string): Promise<Payment> => {
  try {
    const paymentResponse = await formanceSDK.payments.v1.listPayments({
      query: JSON.stringify({
        $match: {
          reference: referenceId,
        },
      }),
    });
    if (paymentResponse.statusCode == 200 && paymentResponse.paymentsCursor.cursor.data.length > 0) {
      return paymentResponse.paymentsCursor.cursor.data[0];
    } else {
      return null;
    }
  } catch (error) {
    logger.error({
      context: "Error finding Formance payment",
      ...error,
    });
    return null;
  }
};

export const updatePaymentMetadata = async (paymentId: string, metadata: Record<string, string>): Promise<void> => {
  try {
    await formanceSDK.payments.v1.updateMetadata({
      requestBody: metadata,
      paymentId,
    });
  } catch (error) {
    logger.error({
      context: "Error updating Formance payment metadata",
      ...error,
    });
  }
};

export const createPayment = async (paymentData: {
  amount: number;
  createdAt: Date;
  reference: string;
  sourceAccountID: string;
  destinationAccountID: string;
}): Promise<string> => {
  try {
    const paymentPayload = {
      amount: BigInt(paymentData.amount),
      asset: "EGP/2",
      connectorID: formanceConnectorId,
      createdAt: paymentData.createdAt,
      reference: paymentData.reference,
      sourceAccountID: paymentData.sourceAccountID,
      destinationAccountID: paymentData.destinationAccountID,
      scheme: PaymentScheme.Other,
      status: PaymentStatus.Succeeded,
      type: PaymentType.PayIn,
    };
    const createPaymentResponse: CreatePaymentResponse = await formanceSDK.payments.v1.createPayment(paymentPayload);
    if (createPaymentResponse.statusCode == 200) {
      return createPaymentResponse.paymentResponse.data.id;
    }
    throw new Error("Error creating Formance payment");
  } catch (error) {
    logger.error({
      context: "Error creating Formance payment",
      ...error,
    });
    return null;
  }
};

export const createAccount = async (consumerId: string): Promise<string> => {
  try {
    const response = await formanceSDK.payments.v1.createAccount({
      accountName: `Internal consumer account for consumer ID ${consumerId}`,
      connectorID: formanceConnectorId,
      createdAt: new Date(),
      defaultAsset: "EGP/2",
      reference: consumerId,
      type: AccountType.Internal,
    });
    if (response.statusCode == 200) {
      return response.paymentsAccountResponse.data.id;
    } else {
      throw Error("Error creating Formance account");
    }
  } catch (error) {
    logger.error({
      context: "Error creating Formance account",
      ...error,
    });
    return null;
  }
};

export const isServiceAvailable = async (): Promise<boolean> => {
  try {
    const response = await formanceSDK.payments.v1.listPayments({});
    return response.statusCode == 200;
  } catch (error) {
    logger.error({
      context: "Error checking Formance health",
      ...error,
    });
    return false;
  }
};

File: legacy/service-department/src/services/formance.py
from datetime import datetime
from config.settings import settings
from src.services.logging import logger
import sdk
from sdk.models import shared, operations
from src.utils.exceptions import ResourceNotCreatedException

formance_base_url = settings.get("formance", default={}).get("base_url", default="")
formance_require_login = settings.get("formance", default={}).get("require_login", default=False)
formance_client_id = settings.get("formance", default={}).get("client_id", default="")
formance_client_secret = settings.get("formance", default={}).get("client_secret", default="")
formance_connector_id = settings.get("formance", default={}).get("connector_id", default="")


logger = logger.bind(service="formance", context="setup", action="setup")


class Formance:
    @staticmethod
    def create_payment(
        amount: int, created_at: datetime, reference: str, destinationAccount: str, metadata: dict = None
    ):
        logger.debug(
            "create_payment with input amount=%s, created_at=%s, reference=%s, destinationAccount=%s, metadata=%s"
            % (amount, created_at, reference, destinationAccount, metadata)
        )
        try:
            # create formance client
            if formance_require_login:
                formance_client = sdk.SDK(
                    server_url=formance_base_url,
                    security=shared.Security(
                        client_id=formance_client_id,
                        client_secret=formance_client_secret,
                    ),
                )
            else:
                formance_client = sdk.SDK(server_url=formance_base_url)

            res = formance_client.payments.create_payment(
                request=shared.PaymentRequest(
                    amount=amount,
                    asset="EGP/2",
                    connector_id=formance_connector_id,
                    created_at=created_at,
                    reference=reference,
                    scheme=shared.PaymentScheme.OTHER,
                    status=shared.PaymentStatus.SUCCEEDED,
                    type=shared.PaymentType.PAY_IN,
                    destination_account_id=destinationAccount,
                )
            )

            if res.payment_response is not None:
                # update the metadata
                metadata_str = {k: str(v) for k, v in metadata.items()}
                formance_client.payments.update_metadata(
                    request=operations.UpdateMetadataRequest(
                        request_body=metadata_str,
                        payment_id=res.payment_response.data.id,
                    )
                )
                return res.payment_response.data
            else:
                raise ResourceNotCreatedException("Payment not created")
        except Exception as exception:
            logger.error("An error occurred: %s" % str(exception))
            raise ResourceNotCreatedException("Payment not created")
